### PR TITLE
Feature/version

### PR DESCRIPTION
**Pull request by KGibb8**
versioning with high level functions that pass to lower level functions

### CODE REVIEW

**Comment by KGibb8** [comment]
@ameba23 @mixmix a code review would be great when you guys get a chance. Here's whats happening:

Each validator takes a message, and finds the relevant schema using the `schemaFinder`. The schemaFinder will take both the name and the message content, so it can check against the version and the schema to see if either are valid.

If we're unsuccessful in location a schema, we pass down a set of errors (that look `like is-my-json-valid` errors so we can tread them in the same way) and return early with `false`.

If we're successful in locating a schema and we have a valid version, then we run the usual set of validations, and pass down any inherited errors.

**Comment by KGibb8** [comment]
I feel like this could be refactored to a point where its super clean and could be used as a standard against all other ssb modules. It'd be nice to do be able to do something like...

```js
// schemas/index.js
const V1_SCHEMA_VERSION = "1"
const V2_SCHEMA_VERSION = "2"

module.exports = {
    V1_SCHEMA_VERSION: require('./v1'),
    V2_SCHEMA_VERSION: require('./v2')
}
```

```js
// isRoot.js
const schemas = require('./schemas')
const validate = require('ssb-validate-schema)

// Returns a function that takes an obj and opts.
module.exports = validate(schemas)

```

**Comment by KGibb8** [comment]
Done a quick refactor into a new module called `ssb-schema-validation` which behaves as above. Haven't tested yet.

https://github.com/ssbc/ssb-schema-valiation

**Comment by ameba23** [comment]
@KGibb8 this looks great.  Really pleased this is happening... 

but i'm having trouble testing, it.  for me there is some abiguity as to how **this** module (meaning 'ssb-dark-crystal-schema') should look in order to utilise your new module.  As in what should the top level 'index.js' contain? should it export 'isRoot.js' which in turn requires './schemas'?  or should it export './schemas' itself?

maybe we should start a new PR for this repo which we can use to test 'ssb-schema-valiation' 

**BTW** there is a typo in the repo name which might cause problems...

**Comment by KGibb8** [comment]
@ameba23 so to get this to run and pass the tests,

```
git clone git@github.com:ssbc/ssb-schema-valiation.git
cd ssb-schema-validation
npm link
cd ../ssb-dark-crystal-schemas
npm link ssb-schema-validation
npm test
```

All tests should pass.

**Comment by ameba23** [comment]
all tests passing my end.  great work!

File: schemas/v2/ritual.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  type: 'object',
+  required: ['type', 'version', 'root', 'quorum', 'shards', 'tool'],
+  properties: {
+    type: {
+      type: 'string',
+      pattern: '^dark-crystal/ritual$'
+    },
+    version: {
+      type: 'string',
+      pattern: `^${SCHEMA_VERSION}$`
```
Comment (nikolai-b) [inline schemas/v2/ritual.js]: Given it isn't really a regex I think the `enum` might make more sense here.
```js
version: {
  type: 'string',
  enum: [SCHEMA_VERSION]
}
```

File: schemas/v2/root.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  type: 'object',
+  required: ['type', 'version', 'name', 'recps'],
+  properties: {
+    type: {
+      type: 'string',
+      pattern: '^dark-crystal/root$'
+    },
+    version: {
+      type: 'string',
+      pattern: `^${SCHEMA_VERSION}$`
```
Comment (nikolai-b) [inline schemas/v2/root.js]: Same here with `enum`

File: schemas/v2/shard.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    version: {
+      type: 'string',
+      pattern: `^${SCHEMA_VERSION}$`
+    },
+    root: { $ref: '#/definitions/messageId' },
+    shard: { $ref: '#/definitions/encrypt/box' },
+    recps: {
+      type: 'array',
+      maxItems: 2,
+      minItems: 2,
```
Comment (nikolai-b) [inline schemas/v2/shard.js]: Would you add `uniqueItems: true` here as presumably that'd stop you sharding to yourself?  Self-sharding seems fine as a test but probably not something you'd want for a real secret.  Jumping in with no context so ignore me if I'm not helping :smile:

Comment (ameba23) [reply schemas/v2/shard.js]: Good point.  Hmm. There might be a situation where giving yourself a shard makes sense.  My opinion is for now we should inhibit this at the interface level, but allow it in the schema.  Others might disagree.

File: schemas/v2/ritual.js (hunk 1 above)
Comment (mixmix) [reply schemas/v2/ritual.js]: I think what's written is correct, unless we do versions which are integer values?
e.g. version `1.0.0` doesn't translate well to enum right?

File: schemas/v2/shard.js (hunk 3 above)
Comment (mixmix) [reply schemas/v2/shard.js]: leave it for the moment IMO
Nice to know that functionality exists though

**Comment by mixmix** [comment]
## Review

Hey @KGibb8 I like to work in this a lot. My only critiques are:
- you left some dead files
- you've added a v2 schema and made it "live" even though it's currently just for testing

I chose to do the tidying work rather than point at thing and you can see what I've done in this PR : 
https://github.com/blockades/ssb-dark-crystal-schema/pull/14

File: schemas/v2/ritual.js (hunk 1 above)
Comment (nikolai-b) [reply schemas/v2/ritual.js]: @mixmix not sure what you mean 
> version `1.0.0` doesn't translate well to enum right?

With the schema:

```json
{
  "$schema": "http://json-schema.org/schema#",
  "type": "object",
  "required": ["version"],
  "properties": {
    "version": {
      "type": "string",
      "enum": ["1.0.0"]
    }
  }
}
```

Then at https://json-schema-validator.herokuapp.com/
```json
{"version": "1.0.1"}
```
fails to validate and
```json
{"version": "1.0.0"}
```
works fine.

Then again they both do the same thing so :woman_shrugging:, just I was looking at all the cool stuff you can do with schema validation :exclamation:

Comment (ameba23) [reply schemas/v2/ritual.js]: @nikolai-b did you come across a way to validate using an arbitrary function? that would be useful for validating returned shards, but i have the feeling its been omitted intentionally.

File: schemas/v2/shard.js (hunk 3 above)
Comment (KGibb8) [reply schemas/v2/shard.js]: Thanks for thoughts Nikolai, more features about is-my-json-valid I didn't know about :)

**Comment by KGibb8** [comment]
@mixmix hey I did remove those files but you must have not pulled the most recent commits. All gravy, I've published to npm and will now merge your forked PR.

**Comment by mixmix** [comment]
weird, don't know what happened there, I usually keep my branches fresh.
anywho !  good work

**Pull request closed without merging**